fhirpipe/transform/fhir.py

```diff
@@ -1,122 +1,199 @@
+def create_instance(row, mapping):
+    """ Function used to create a single FHIR instance.
+    """
+    # Modify the data structure so that it is easier to use
+    path_attributes_map = {attr["path"]: attr for attr in mapping["attributes"]}
 
-    return fhir_object
+    # Build path value map
+    fhir_object = build_fhir_object(row, path_attributes_map)
 
+    # Identify the fhir object
+    fhir_object["id"] = str(uuid4())
+    fhir_object["resourceType"] = mapping["definition"]["type"]
 
-def build_fhir_leaf_attribute(structure, row, indices):
-    # TODO simplify (for instance, force merging script when several inputs)
-    cols_to_fetch = ([], [])
-    for inp in structure["inputs"]:
-
-        # If a sql location is provided, then a sql value has been returned
-        if inp["sqlValue"]:
-            sql = inp["sqlValue"]
-            col_name = build_col_name(sql["table"], sql["column"], sql["owner"])
-            if inp["script"]:
-                # If there is a cleaning script, we fetch the value in the cleaned column
-                col_name = new_col_name(inp["script"], col_name)
-            cols_to_fetch[0].append(col_name)
+    # Remove duplicates in fhir object
+    fhir_object = clean_fhir_object(fhir_object)
 
-        # Else retrieve the static value
-        else:
-            cols_to_fetch[1].append(inp["staticValue"])
+    return fhir_object
 
-    if structure["mergingScript"]:
-        result = row[new_col_name(structure["mergingScript"], cols_to_fetch)]
-    else:
-        result = [row[c] for c in cols_to_fetch[0]]
-        result.extend(cols_to_fetch[1])
 
-    # Unlist
-    if isinstance(result, list) and len(result) == 1:
-        result = result[0]
+def build_fhir_object(row, path_attributes_map, index=None):
+    """ Function that actually builds a nested object from the dataframe row and the mapping.
+    Note that it can be used to build only a subpart of a fhir instance.
+    """
+    fhir_object = recursive_defaultdict()
+    arrays_done = set()
 
-    if isinstance(result, tuple):
-        for ind in indices:
-            result = result[ind]
+    for path, attr in path_attributes_map.items():
+        if attr["inputs"]:
 
-    if isinstance(result, tuple):
-        if not all([el == result[0] for el in result[1:]]):
-            raise Exception(f"Cannot have as input a list with different values ({result}).")
-        result = result[0]
+            split_path = path.split(".")
+            position_ind = get_position_first_index(split_path)
 
-    return result
+            if position_ind is None:
+                val = fetch_values_from_dataframe(row, attr["inputs"], attr["mergingScript"])
+                if isinstance(val, tuple) and len(val) == 1:
+                    insert_in_fhir_object(fhir_object, path, val[0])
+                elif isinstance(val, tuple) and index is not None:
+                    insert_in_fhir_object(fhir_object, path, val[index])
```

Comment:
Is the case `isinstance(val, tuple) and len(val) != 1 and index is None` an error ?
If so I'd rather write:
```python
if isinstance(val, tuple):
    if len(val) == 1:
         ...
    elif index is not None:
         ...
    else:
        raise MyError()
else:
```

Reply:
It's not necessarily an error if all the values are the same (comment to come line ~150)